### PR TITLE
Dataset APIs - data ingestion

### PR DESCRIPTION
**Pull request by ly-kolena**
### Linked issue(s):

Example codes converted from object_detection_2d example script

```
def load_transportation_data() -> Dict[str, List[LabeledBoundingBox]]:
    # same codes from example
    ...


def complete_datapoints(args: Namespace) -> pd.DataFrame:
    print("loading S3 files...")
    image_to_boxes = load_transportation_data()
    data = pd.read_csv(args.metadata)
    dataset = data[["width", "height", "brightness"]]
    dataset["locator"] = S3_IMAGE_LOCATION + data["name"]
    dataset["bboxes"] = data["name"].apply(lambda x: image_to_boxes[x])

    return dataset


def main(args: Namespace) -> None:
    kolena.initialize(os.environ["KOLENA_TOKEN"], verbose=True)

    register_dataset(DATASET, complete_datapoints(args))
    df_loaded = fetch_dataset(DATASET)
    print(df_loaded.shape)


if __name__ == "__main__":
    ap = ArgumentParser()
    ap.add_argument(
        "--metadata",
        type=str,
        default="s3://kolena-public-datasets/coco-2014-val/meta/metadata_attribution2.0_brightness.csv",
        help="The path to the image metadata stored in a csv.",
    )

    main(ap.parse_args())
```

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

### CODE REVIEW

**Comment by ly-kolena** [comment]
cc @gordonhart @munkyshi for early review

File: kolena/_experimental/dataset/dataset.py
Comment (ly-kolena) [inline kolena/_experimental/dataset/dataset.py]: Any suggestion better than `dataset.dataset`?

File: kolena/workflow/_datatypes.py
Comment (ly-kolena) [inline kolena/workflow/_datatypes.py]: Should be gone after https://github.com/kolenaIO/kolena/pull/239 is merged

File: tests/unit/workflow/test_data_object.py
Comment (ly-kolena) [inline tests/unit/workflow/test_data_object.py]: Should be gone after #239 is merged

**Comment by y27choi** [comment]
Do we want to return the dataset object from `register_dataset(DATASET, complete_datapoints(args))`?

File: kolena/_experimental/dataset/dataset.py
Comment (y27choi) [reply kolena/_experimental/dataset/dataset.py]: can we add `register_dataset()` and `fetch_dataset()` to `__init__.py` so we can import them by `from kolena._experimental.dataset import register_dataset, fetch_dataset`?

**Comment by ly-kolena** [comment]
> Do we want to return the dataset object from `register_dataset(DATASET, complete_datapoints(args))`?

Since we tried to avoid introducing a new `Dataset` object, and went with operating on `DataFrame`, the server response was not passed back to users in this implementation. Users currently cannot do anything with this return. Do you think it'd be useful if a plain `Dict` is returned back from `register_dataset`?

File: kolena/_experimental/dataset/dataset.py
Comment (ly-kolena) [reply kolena/_experimental/dataset/dataset.py]: updated as suggested

**Comment by y27choi** [comment]
> Do you think it'd be useful if a plain Dict is returned back from register_dataset? 

Probably not. I was thinking that I'd need the dataset object to run tests but I just need the dataset name to kick off a run, right? In that case, I don't think I would actually need the dataset object for the happy path. In which scenarios do you think I would ever need the object?

**Comment by ly-kolena** [comment]
> Probably not. I was thinking that I'd need the dataset object to run tests but I just need the dataset name to kick off a run, right? In that case, I don't think I would actually need the dataset object for the happy path. In which scenarios do you think I would ever need the object?

Current thinking is (hopefully) users can do all operations with minimal info needed, so dataset name (and model name) would be all they need in the most common cases to send data to kolena.

**Comment by y27choi** [comment]
And things like tags, description are something users would need to add from the frontend?

**Comment by ly-kolena** [comment]
> And things like tags, description are something users would need to add from the frontend?

That is the expectation thus far.

**Comment by y27choi** [comment]
The interface for dataset looks very simple which I like.  I will play around with different annotation types / cell types later but for now it's 👍 from me.

File: kolena/_experimental/dataset/_dataset.py
Comment (munkyshi) [inline kolena/_experimental/dataset/_dataset.py:129]: I would expect `iter_dataset` to iterate over datapoints, not over data frames -- IMO it's an awkward interface with data frames.

**Comment by munkyshi** [comment]
> > And things like tags, description are something users would need to add from the frontend?
> 
> That is the expectation thus far.

I would expect programmatic tags to be useful, especially if we're using tags in place of workflows (to indicate which remote evaluators are compatible)

**Review by munkyshi — APPROVED** [review]
I'd like to reach an agreement on https://github.com/kolenaIO/kolena/pull/201/files/5b4b94915192dd7732504b69d19e1647ba43a504#diff-2b897e1ce619a688e8b9627675e7d0f5509b44518717c29b2cb7755a9b7993f8 before merging, but everything else looks good. Approving to not block on merge after this gets sorted out.

**Review by munkyshi — APPROVED** [review]
Synced with @ly-kolena offline -- I'm convinced to move forward here as is. We can work through programmatic tags separately in the future, depending on validation on this approach from @y27choi's prototyping.

File: kolena/_experimental/dataset/_dataset.py
Comment (dylangrandmont) [inline kolena/_experimental/dataset/_dataset.py]: nit: It would make sense to declare the enum of possible types directly and reference that here for the map values. We do something similar in `marina` (with typescript):

```ts
export enum TestSampleType {
  IMAGE = 'TEST_SAMPLE/IMAGE',
  COMPOSITE = 'TEST_SAMPLE/COMPOSITE',
  TEXT = 'TEST_SAMPLE/TEXT',
  VIDEO = 'TEST_SAMPLE/VIDEO',
  DOCUMENT = 'TEST_SAMPLE/DOCUMENT',
  POINT_CLOUD = 'TEST_SAMPLE/POINT_CLOUD',
}
```

Comment (dylangrandmont) [reply kolena/_experimental/dataset/_dataset.py]: Likewise below, for instance [here](https://github.com/kolenaIO/kolena/pull/201/files#diff-2b897e1ce619a688e8b9627675e7d0f5509b44518717c29b2cb7755a9b7993f8R67).

File: tests/unit/_experimental/dataset/test_dataset.py
Comment (dylangrandmont) [inline tests/unit/_experimental/dataset/test_dataset.py]: [Declaring an enum](https://github.com/kolenaIO/kolena/pull/201/files#r1337314296) would also remove the need for this to be declared in a test file

File: kolena/_experimental/dataset/_dataset.py
Comment (dylangrandmont) [inline kolena/_experimental/dataset/_dataset.py:77]: Is this not covered by `mimetypes.guess_type`?

File: kolena/_experimental/dataset/_dataset.py
Comment (dylangrandmont) [inline kolena/_experimental/dataset/_dataset.py]: Potentially this should map to `TEST_SAMPLE/TEXT`, which `marina` already supports. Which mime types will this include?

Comment (ly-kolena) [reply kolena/_experimental/dataset/_dataset.py]: How does marina handle a test sample with `locator: https://example.com/test.xml`? The `text` type here is for this case, not a test sample using `text: "my qa question"`.

File: kolena/_experimental/dataset/_dataset.py
Comment (ly-kolena) [reply kolena/_experimental/dataset/_dataset.py:77]: It does not. For reference, [mimetype](https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type) package references [this list ](https://www.iana.org/assignments/media-types/media-types.xhtml)

File: kolena/_experimental/dataset/_dataset.py
Comment (ly-kolena) [reply kolena/_experimental/dataset/_dataset.py]: updated as suggested

**Review by yifanwu-kolena — APPROVED** [review]

File: kolena/_experimental/dataset/_dataset.py
Comment (dylangrandmont) [inline kolena/_experimental/dataset/_dataset.py:44]: We should avoid all use of the terminology "test sample" and instead be using "datapoint"

**Pull request merged**
